presto-hive/src/test/java/com/facebook/presto/hive/TestHiveRecoverableGroupedExecution.java

```diff
@@ -0,0 +1,246 @@
+                queryRunner.execute(recoverableSession, "DROP TABLE IF EXISTS test_insert");
+            }
+        }
+    }
+
+    @Test(timeOut = 60_000)
+    public void testRecoverableGroupedExecutionWithFailure()
+            throws Exception
+    {
+        try (DistributedQueryRunner queryRunner = distributedQueryRunnerSupplier.get()) {
+            try {
+                queryRunner.execute(
+                        recoverableSession,
+                        "CREATE TABLE test_table1\n" +
+                                "WITH (bucket_count = 13, bucketed_by = ARRAY['key1']) AS\n" +
+                                "SELECT orderkey key1, comment value1 FROM orders");
+                queryRunner.execute(
+                        recoverableSession,
+                        "CREATE TABLE test_table2\n" +
+                                "WITH (bucket_count = 13, bucketed_by = ARRAY['key2']) AS\n" +
+                                "SELECT orderkey key2, comment value2 FROM orders");
+                queryRunner.execute(
+                        recoverableSession,
+                        "CREATE TABLE test_table3\n" +
+                                "WITH (bucket_count = 13, bucketed_by = ARRAY['key3']) AS\n" +
+                                "SELECT orderkey key3, comment value3 FROM orders");
+                queryRunner.execute(
+                        recoverableSession,
+                        "CREATE TABLE test_insert (key BIGINT, value VARCHAR, partition_key VARCHAR)\n" +
+                                "WITH (bucket_count = 13, bucketed_by = ARRAY['key'], partitioned_by = ARRAY['partition_key'])");
+
+                ListenableFuture<MaterializedResult> createTable = executor.submit(() -> queryRunner.execute(
+                        recoverableSession,
+                        "CREATE TABLE test_create_table WITH (bucket_count = 13, bucketed_by = ARRAY['key1']) AS\n" +
+                                "SELECT key1, value1, key2, value2, key3, value3\n" +
+                                "FROM test_table1\n" +
+                                "JOIN test_table2\n" +
+                                "ON key1 = key2\n" +
+                                "JOIN test_table3\n" +
+                                "ON key2 = key3"));
+                ListenableFuture<MaterializedResult> insert = executor.submit(() -> queryRunner.execute(
+                        recoverableSession,
+                        "INSERT INTO test_insert\n" +
+                                "SELECT key1, value1, 'foo'\n" +
+                                "FROM test_table1\n" +
+                                "JOIN test_table2\n" +
+                                "ON key1 = key2\n" +
+                                "JOIN test_table3\n" +
+                                "ON key2 = key3"));
+
+                while (getRunningQueryCount(queryRunner.getQueries()) != 2) {
```

Comment:
Add some comments here and around line 195, like "wait queries start to run", "wait for additional 500ms before close a worker.